### PR TITLE
Jv homework fruit shop v.2

### PR DESCRIPTION
**Pull request by etezaadi**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/CsvFileReaderService.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface CsvFileReaderService {
+    List<String> fileReader(String fileName);
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/CsvFileReaderService.java:6]:
```suggestion
    List<String> readFromFile(String fileName);
```

File: src/main/java/core/basesyntax/service/CsvFileWriterService.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.service;
+
+public interface CsvFileWriterService {
+    void fileWriter(String fileName);
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/CsvFileWriterService.java:4]:
```suggestion
    void writeToFile(String fileName);
```

File: src/main/java/core/basesyntax/service/FruitTransaction.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface FruitTransaction {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/FruitTransaction.java:5]:
```suggestion
public interface FruitTransactionService {
```

File: src/main/java/core/basesyntax/service/impl/CsvFileReaderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.service.CsvFileReaderService;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+
+public class CsvFileReaderServiceImpl implements CsvFileReaderService {
+
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/CsvFileReaderServiceImpl.java:10]: remove empty line

File: src/main/java/core/basesyntax/service/impl/CsvFileWriterServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class CsvFileWriterServiceImpl implements CsvFileWriterService {
+    public static final String COLUMN = "fruit,quantity";
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/CsvFileWriterServiceImpl.java:14]:
```suggestion
    private static final String HEADER = "fruit,quantity";
```

File: src/main/java/core/basesyntax/service/impl/CsvFileWriterServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        List<String> fromMapToList = new ArrayList<>();
+        for (Map.Entry<String, Integer> output : Storage.fruitStorage.entrySet()) {
+            builder.append(output.getKey())
+                    .append(",")
+                    .append(output.getValue())
+                    .append(System.lineSeparator());
+            fromMapToList.add(builder.toString());
+        }
+        return fromMapToList;
+    }
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/CsvFileWriterServiceImpl.java:41]: it's better to create a separate service that will form a report

File: src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class FruitTransactionImpl implements FruitTransaction {
+    private static final int INDEX_OPERATION = 0;
+    private static final int INDEX_NAME = 1;
+    private static final int START_FIND_ELEMENTS = 1;
+    private static final int INITIALIZATION_QUANTITY_IN_DB = 0;
+    private static final String SPLIT_CHARACTER = ",";
+    private static FruitShopDao fruitShopDao = new FruitShopDaoImpl();
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java:17]: let's initialize this field using constructor

File: src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.service.FruitTransaction;
+import core.basesyntax.service.strategy.OperationStrategy;
+import java.util.ArrayList;
+import java.util.List;
+
+public class FruitTransactionImpl implements FruitTransaction {
+    private static final int INDEX_OPERATION = 0;
+    private static final int INDEX_NAME = 1;
+    private static final int START_FIND_ELEMENTS = 1;
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java:14]:
```suggestion
    private static final int LINE_OF_FIRST_OPERATION = 1;
```

File: src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import core.basesyntax.dao.FruitShopDaoImpl;
+import core.basesyntax.db.Storage;
+import core.basesyntax.service.FruitTransaction;
+import core.basesyntax.service.strategy.OperationStrategy;
+import java.util.ArrayList;
+import java.util.List;
+
+public class FruitTransactionImpl implements FruitTransaction {
+    private static final int INDEX_OPERATION = 0;
+    private static final int INDEX_NAME = 1;
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java:13]:
```suggestion
    private static final int OPERATION_INDEX = 0;
    private static final int NAME_NAME = 1;
```

File: src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    private static final int START_FIND_ELEMENTS = 1;
+    private static final int INITIALIZATION_QUANTITY_IN_DB = 0;
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java:15]: it's a redundant constant, remove it

File: src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    private static FruitShopDao fruitShopDao = new FruitShopDaoImpl();
+    private OperationStrategy operationStrategy;
+
+    public FruitTransactionImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void dateProcessing(List<String> data) {
+        validateDB(data);
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java:26]: let's rewrite your logic, so you can remove this method from this class, you don't need to initialize fruits before processing each operation

File: src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        for (int i = START_FIND_ELEMENTS; i < data.size(); i++) {
+            String[] splitElement = data.get(i).split(SPLIT_CHARACTER);
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/FruitTransactionImpl.java:28]:
```suggestion
            String[] splittedLine = lines.get(i).split(SPLIT_CHARACTER);
```

File: src/main/java/core/basesyntax/service/strategy/OperationHandler.java [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.service.strategy;
+
+public interface OperationHandler {
+    void getOperation(String[] data);
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/strategy/OperationHandler.java:4]:
```suggestion
    void handle(String fruitName, int quantity);
```

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.service.SeparateService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class SeparateServiceImpl implements SeparateService {
+    public static final String HEADER = "fruit,quantity";
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java:10]:
```suggestion
    private static final String HEADER = "fruit,quantity";
```

File: src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Override
+    public List<String> getReportFromList() {
+        StringBuilder builder = new StringBuilder();
+        builder.append(HEADER)
+                .append(System.lineSeparator());
+        List<String> fromMapToList = new ArrayList<>();
+        for (Map.Entry<String, Integer> output : Storage.fruitStorage.entrySet()) {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java:18]: let's create a method getAll in FruitShopDao that returns `Storage.fruitStorage`;

File: src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+public class SeparateServiceImpl implements SeparateService {
+    public static final String HEADER = "fruit,quantity";
+
+    @Override
+    public List<String> getReportFromList() {
+        StringBuilder builder = new StringBuilder();
+        builder.append(HEADER)
+                .append(System.lineSeparator());
+        List<String> fromMapToList = new ArrayList<>();
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java:17]: you don't this list, remove it

File: src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+import core.basesyntax.service.SeparateService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class SeparateServiceImpl implements SeparateService {
+    public static final String HEADER = "fruit,quantity";
+
+    @Override
+    public List<String> getReportFromList() {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/SeparateServiceImpl.java:13]:
```suggestion
    public String getReportFromList() {
```

File: src/main/java/core/basesyntax/service/SeparateService.java [hunk 18]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface SeparateService {
+    List<String> getReportFromList();
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/SeparateService.java:6]:
```suggestion
    String createReport();
```

File: src/main/java/core/basesyntax/service/SeparateService.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface SeparateService {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/SeparateService.java:5]:
```suggestion
public interface ReportService {
```

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/impl/ReportServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+import core.basesyntax.service.ReportService;
+import java.util.Map;
+
+public class ReportServiceImpl implements ReportService {
+    private static final String HEADER = "fruit,quantity";
+    private FruitShopDao fruitShopDao = new FruitShopDaoImpl();
+
+    @Override
+    public String createReport() {
+        StringBuilder builder = new StringBuilder();
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/ReportServiceImpl.java:14]:
```suggestion
        StringBuilder reportBuilder = new StringBuilder();
```

File: src/main/java/core/basesyntax/service/impl/ReportServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,25 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.dao.FruitShopDao;
+import core.basesyntax.dao.FruitShopDaoImpl;
+import core.basesyntax.service.ReportService;
+import java.util.Map;
+
+public class ReportServiceImpl implements ReportService {
+    private static final String HEADER = "fruit,quantity";
+    private FruitShopDao fruitShopDao = new FruitShopDaoImpl();
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/impl/ReportServiceImpl.java:10]: initialize this field, using constructor

**Review by Nick97-git — CHANGES_REQUESTED** [review]

**Comment by etezaadi** [comment]
@Nick97-git i thank you for your help :)

**Review by Nick97-git — APPROVED** [review]

**Pull request closed without merging**